src/cpp/alohalytics.cc

```diff
@@ -46,172 +49,86 @@
-  assert(!url.empty());
-  HTTPClientPlatformWrapper request(url);
-  request.set_debug_mode(debug_mode);
-
-  try {
-    // TODO(AlexZ): Refactor FileStorageQueue to automatically append ID and gzip files, so we don't need
-    // temporary memory buffer any more and files take less space.
-    request.set_body_data(alohalytics::Gzip(buffer), "application/alohalytics-binary-blob", "POST", "gzip");
-    return request.RunHTTPRequest() && 200 == request.error_code() && !request.was_redirected();
-  } catch (const std::exception& ex) {
-    if (debug_mode) {
-      ALOG("Exception while trying to UploadBuffer", ex.what());
-    }
-  }
-  return false;
-}
+Stats::Stats()
+    : messages_queue_(
+          std::bind(&Stats::GzipAndArchiveFileInTheQueue, this, std::placeholders::_1, std::placeholders::_2)) {}
 
-// Processes messages passed from UI in message queue's own thread.
-// TODO(AlexZ): Refactor message queue to make this method private.
-void Stats::OnMessage(const MQMessage& message, size_t dropped_events) {
-  if (dropped_events) {
-    LOG_IF_DEBUG("Warning:", dropped_events, "events were dropped from the queue.");
+void Stats::GzipAndArchiveFileInTheQueue(const std::string & in_file, const std::string & out_archive) {
+  if (unique_client_id_event_.empty()) {
+    LOG_IF_DEBUG("Warning: unique client id is not set in GzipAndArchiveFileInTheQueue.")
+  } else {
+    LOG_IF_DEBUG("Archiving", in_file, "to", out_archive);
   }
-
-  if (message.ForceUpload()) {
-    if (upload_url_.empty()) {
-      LOG_IF_DEBUG("Warning: Can't upload in-memory events because upload url has not been set.");
-      return;
-    }
-    LOG_IF_DEBUG("Forcing statistics uploading.");
-    if (file_storage_queue_) {
-      // Upload all data, including 'current' file.
-      file_storage_queue_->ForceProcessing(true);
-    } else {
-      std::string buffer = unique_client_id_event_;
-      // TODO(AlexZ): thread-safety?
-      TMemoryContainer copy;
-      copy.swap(memory_storage_);
-      for (const auto& evt : copy) {
-        buffer.append(evt);
-      }
-      LOG_IF_DEBUG("Forcing in-memory statistics uploading.");
-      if (!UploadBuffer(upload_url_, std::move(buffer), debug_mode_)) {
-        // If failed, merge events we tried to upload with possible new ones.
-        memory_storage_.splice(memory_storage_.end(), std::move(copy));
-        LOG_IF_DEBUG("Failed to upload in-memory statistics.");
-      }
+  // Append unique installation id in the beginning of each archived file.
+  // If it is empty then all stats data will become completely anonymous.
+  try {
+    std::string buffer(unique_client_id_event_);
+    {
+      std::ifstream fi;
+      fi.exceptions(std::ifstream::failbit | std::ifstream::badbit);
+      fi.open(in_file, std::ifstream::in | std::ifstream::binary);
+      const size_t id_size = unique_client_id_event_.size();
+      const int64_t file_size = FileManager::GetFileSize(in_file);
```

Comment:
Which you then `static_cast` to an unsigned type in production -- proving it's a bad design for the 21st century.

Throw an exception. Or have it return `bool` and initialize an output parameter with size. The way you did it only creates more room for error, at questionable cost.